### ISSUE TITLE
Building with a macOS plugin: "Plugin `url_launcher_linux` doesn't implement a plugin interface"

### ISSUE THREAD

**Issue by jmagman** [flutter/flutter#76396]
On top of #74469 I see confusing output when I build for a macOS app
```
Plugin `url_launcher_linux` doesn't implement a plugin interface, nor sets a default implementation in pubspec.yaml.

To set a default implementation, use:
flutter:
  plugin:
    platforms:
      linux:
        default_package: <plugin-implementation>

To implement an interface, use:
flutter:
  plugin:
    implements: <plugin-interface>

Plugin `url_launcher_macos` doesn't implement a plugin interface, nor sets a default implementation in pubspec.yaml.

To set a default implementation, use:
flutter:
  plugin:
    platforms:
      macos:
        default_package: <plugin-implementation>

To implement an interface, use:
flutter:
  plugin:
    implements: <plugin-interface>

Plugin `url_launcher_windows` doesn't implement a plugin interface, nor sets a default implementation in pubspec.yaml.
To set a default implementation, use:
flutter:
  plugin:
    platforms:
      windows:
        default_package: <plugin-implementation>

To implement an interface, use:

flutter:
  plugin:
    implements: <plugin-interface>
```

### Steps to Reproduce
1. flutter create --platforms=macos test_macos_plugin
2. Add `url_launcher:` to pubspec.yaml
3. Run in Xcode or `flutter run -v`

**Comment by jmagman** [comment]
@blasten

**Comment by blasten** [comment]
I was unable to reproduce it. what version did it resolve to (from .packages)?

**Comment by blasten** [comment]
Actually, I had modified the pub cache in order to test the changes.  This is the intended behavior, and I will submit a fix to the offending plugins.

**Comment by CodinMoldovanu** [comment]
@jmagman What Flutter version are you using? I was encountering this while using the `master` branch of the Flutter repo, switched to stable and everything is behaving nicely.

**Comment by narumishi** [comment]
This should be warnings for `plugin` which not implements all platforms, should not shown when building `application`?

```
Plugin `audioplayers` doesn't implement a plugin interface, nor sets a default implementation in pubspec.yaml.
                                                                        
To set a default implementation, use:                                   
flutter:                                                                
  plugin:                                                               
    platforms:                                                          
      macos:                                                            
        default_package: <plugin-implementation>                        
                                                                        
To implement an interface, use:                                         
flutter:                                                                
  plugin:                                                               
    implements: <plugin-interface>                                      
                                                                        
Plugin `connectivity_plus_linux` doesn't implement a plugin interface, nor sets a default implementation in pubspec.yaml.
                                                                        
To set a default implementation, use:                                   
flutter:                                                                
  plugin:                                                               
    platforms:                                                          
      linux:                                                            
        default_package: <plugin-implementation>                        
                                                                        
To implement an interface, use:                                         
flutter:                                                                
  plugin:                                                               
    implements: <plugin-interface>                                      
                                                                        
Plugin `connectivity_plus_macos` doesn't implement a plugin interface, nor sets a default implementation in pubspec.yaml.
                                                                        
To set a default implementation, use:                                   
flutter:                                                                
  plugin:                                                               
    platforms:                                                          
      macos:                                                            
        default_package: <plugin-implementation>                        
                                                                        
To implement an interface, use:                                         
flutter:                                                                
  plugin:                                                               
    implements: <plugin-interface>                                      
                                                                        
Plugin `connectivity_plus_windows` doesn't implement a plugin interface, nor sets a default implementation in pubspec.yaml.
                                                                        
To set a default implementation, use:                                   
flutter:                                                                
  plugin:                                                               
    platforms:                                                          
      windows:                                                          
        default_package: <plugin-implementation>                        
                                                                        
To implement an interface, use:                                         
flutter:                                                                
  plugin:                                                               
    implements: <plugin-interface>                                      
                                                                        
Plugin `file_chooser` doesn't implement a plugin interface, nor sets a default implementation in pubspec.yaml.
                                                                        
To set a default implementation, use:                                   
flutter:                                                                
  plugin:                                                               
    platforms:                                                          
      linux:                                                            
        default_package: <plugin-implementation>                        
                                                                        
To implement an interface, use:                                         
flutter:                                                                
  plugin:                                                               
    implements: <plugin-interface>                                      
                                                                        
Plugin `file_chooser` doesn't implement a plugin interface, nor sets a default implementation in pubspec.yaml.
                                                                        
To set a default implementation, use:                                   
flutter:                                                                
  plugin:                                                               
    platforms:                                                          
      macos:                                                            
        default_package: <plugin-implementation>                        
                                                                        
To implement an interface, use:                                         
flutter:                                                                
  plugin:                                                               
    implements: <plugin-interface>                                      
                                                                        
Plugin `file_chooser` doesn't implement a plugin interface, nor sets a default implementation in pubspec.yaml.
                                                                        
To set a default implementation, use:                                   
flutter:                                                                
  plugin:                                                               
    platforms:                                                          
      windows:                                                          
        default_package: <plugin-implementation>                        
                                                                        
To implement an interface, use:                                         
flutter:                                                                
  plugin:                                                               
    implements: <plugin-interface>                                      
                                                                        
Plugin `flutter_audio_desktop` doesn't implement a plugin interface, nor sets a default implementation in pubspec.yaml.
                                                                        
To set a default implementation, use:                                   
flutter:                                                                
  plugin:                                                               
    platforms:                                                          
      linux:                                                            
        default_package: <plugin-implementation>                        
                                                                        
To implement an interface, use:                                         
flutter:                                                                
  plugin:                                                               
    implements: <plugin-interface>                                      
                                                                        
Plugin `flutter_audio_desktop` doesn't implement a plugin interface, nor sets a default implementation in pubspec.yaml.
                                                                        
To set a default implementation, use:                                   
flutter:                                                                
  plugin:                                                               
    platforms:                                                          
      windows:                                                          
        default_package: <plugin-implementation>                        
                                                                        
To implement an interface, use:                                         
flutter:                                                                
  plugin:                                                               
    implements: <plugin-interface>                                      
                                                                        
Plugin `flutter_qjs` doesn't implement a plugin interface, nor sets a default implementation in pubspec.yaml.
                                                                        
To set a default implementation, use:                                   
flutter:                                                                
  plugin:                                                               
    platforms:                                                          
      linux:                                                            
        default_package: <plugin-implementation>                        
                                                                        
To implement an interface, use:                                         
flutter:                                                                
  plugin:                                                               
    implements: <plugin-interface>                                      
                                                                        
Plugin `flutter_qjs` doesn't implement a plugin interface, nor sets a default implementation in pubspec.yaml.
                                                                        
To set a default implementation, use:                                   
flutter:                                                                
  plugin:                                                               
    platforms:                                                          
      macos:                                                            
        default_package: <plugin-implementation>                        
                                                                        
To implement an interface, use:                                         
flutter:                                                                
  plugin:                                                               
    implements: <plugin-interface>                                      
                                                                        
Plugin `flutter_qjs` doesn't implement a plugin interface, nor sets a default implementation in pubspec.yaml.
                                                                        
To set a default implementation, use:                                   
flutter:                                                                
  plugin:                                                               
    platforms:                                                          
      windows:                                                          
        default_package: <plugin-implementation>                        
                                                                        
To implement an interface, use:                                         
flutter:                                                                
  plugin:                                                               
    implements: <plugin-interface>                                      
                                                                        
Plugin `package_info` doesn't implement a plugin interface, nor sets a default implementation in pubspec.yaml.
                                                                        
To set a default implementation, use:                                   
flutter:                                                                
  plugin:                                                               
    platforms:                                                          
      macos:                                                            
        default_package: <plugin-implementation>                        
                                                                        
To implement an interface, use:                                         
flutter:                                                                
  plugin:                                                               
    implements: <plugin-interface>                                      
                                                                        
Plugin `path_provider_linux` doesn't implement a plugin interface, nor sets a default implementation in pubspec.yaml.
                                                                        
To set a default implementation, use:                                   
flutter:                                                                
  plugin:                                                               
    platforms:                                                          
      linux:                                                            
        default_package: <plugin-implementation>                        
                                                                        
To implement an interface, use:                                         
flutter:                                                                
  plugin:                                                               
    implements: <plugin-interface>                                      
                                                                        
Plugin `path_provider_macos` doesn't implement a plugin interface, nor sets a default implementation in pubspec.yaml.
                                                                        
To set a default implementation, use:                                   
flutter:                                                                
  plugin:                                                               
    platforms:                                                          
      macos:                                                            
        default_package: <plugin-implementation>                        
                                                                        
To implement an interface, use:                                         
flutter:                                                                
  plugin:                                                               
    implements: <plugin-interface>                                      
                                                                        
Plugin `path_provider_windows` doesn't implement a plugin interface, nor sets a default implementation in pubspec.yaml.
                                                                        
To set a default implementation, use:                                   
flutter:                                                                
  plugin:                                                               
    platforms:                                                          
      windows:                                                          
        default_package: <plugin-implementation>                        
                                                                        
To implement an interface, use:                                         
flutter:                                                                
  plugin:                                                               
    implements: <plugin-interface>                                      
                                                                        
Plugin `shared_preferences_linux` doesn't implement a plugin interface, nor sets a default implementation in pubspec.yaml.
                                                                        
To set a default implementation, use:                                   
flutter:                                                                
  plugin:                                                               
    platforms:                                                          
      linux:                                                            
        default_package: <plugin-implementation>                        
                                                                        
To implement an interface, use:                                         
flutter:                                                                
  plugin:                                                               
    implements: <plugin-interface>                                      
                                                                        
Plugin `shared_preferences_macos` doesn't implement a plugin interface, nor sets a default implementation in pubspec.yaml.
                                                                        
To set a default implementation, use:                                   
flutter:                                                                
  plugin:                                                               
    platforms:                                                          
      macos:                                                            
        default_package: <plugin-implementation>                        
                                                                        
To implement an interface, use:                                         
flutter:                                                                
  plugin:                                                               
    implements: <plugin-interface>                                      
                                                                        
Plugin `shared_preferences_windows` doesn't implement a plugin interface, nor sets a default implementation in pubspec.yaml.
                                                                        
To set a default implementation, use:                                   
flutter:                                                                
  plugin:                                                               
    platforms:                                                          
      windows:                                                          
        default_package: <plugin-implementation>                        
                                                                        
To implement an interface, use:                                         
flutter:                                                                
  plugin:                                                               
    implements: <plugin-interface>                                      
                                                                        
Plugin `sqflite` doesn't implement a plugin interface, nor sets a default implementation in pubspec.yaml.
                                                                        
To set a default implementation, use:                                   
flutter:                                                                
  plugin:                                                               
    platforms:                                                          
      macos:                                                            
        default_package: <plugin-implementation>                        
                                                                        
To implement an interface, use:                                         
flutter:                                                                
  plugin:                                                               
    implements: <plugin-interface>                                      
                                                                        
Plugin `url_launcher_linux` doesn't implement a plugin interface, nor sets a default implementation in pubspec.yaml.
                                                                        
To set a default implementation, use:                                   
flutter:                                                                
  plugin:                                                               
    platforms:                                                          
      linux:                                                            
        default_package: <plugin-implementation>                        
                                                                        
To implement an interface, use:                                         
flutter:                                                                
  plugin:                                                               
    implements: <plugin-interface>                                      
                                                                        
Plugin `url_launcher_macos` doesn't implement a plugin interface, nor sets a default implementation in pubspec.yaml.
                                                                        
To set a default implementation, use:                                   
flutter:                                                                
  plugin:                                                               
    platforms:                                                          
      macos:                                                            
        default_package: <plugin-implementation>                        
                                                                        
To implement an interface, use:                                         
flutter:                                                                
  plugin:                                                               
    implements: <plugin-interface>                                      
                                                                        
Plugin `url_launcher_windows` doesn't implement a plugin interface, nor sets a default implementation in pubspec.yaml.
                                                                        
To set a default implementation, use:                                   
flutter:                                                                
  plugin:                                                               
    platforms:                                                          
      windows:                                                          
        default_package: <plugin-implementation>                        
                                                                        
To implement an interface, use:                                         
flutter:                                                                
  plugin:                                                               
    implements: <plugin-interface>   
```

**Comment by jmagman** [comment]
> @jmagman What Flutter version are you using? I was encountering this while using the `master` branch of the Flutter repo, switched to stable and everything is behaving nicely.

I was on https://github.com/flutter/flutter/pull/74469 which is master.

@blasten Upping the priority of this.  Instead of just fixing the offending plugins, can we suppress this warning in tooling when building an app?  The plugin consumer can't do anything about this warning and it looks scary.

**Comment by blasten** [comment]
Yes. I was going to disable the warnings for the time being.

**Comment by Mhamza-MM** [comment]
Having same issue

`FAILURE: Build failed with an exception.

* Where:
Script '/Users/macbook/development/flutter/packages/flutter_tools/gradle/flutter.gradle' line: 991

* What went wrong:
Execution failed for task ':app:compileFlutterBuildDebug'.
> Process 'command '/Users/macbook/development/flutter/bin/flutter'' finished with non-zero exit value 1

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 43s
Exception: Gradle task assembleDebug failed with exit code 1
`

flutter doctor ouput: 

`Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel master, 1.27.0-5.0.pre.92, on macOS 11.2 20D64 darwin-x64, locale en-PK)
[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
[✓] Xcode - develop for iOS and macOS
[✓] Chrome - develop for the web
[✓] Android Studio (version 4.1)
[✓] IntelliJ IDEA Ultimate Edition (version 2020.1)
[✓] VS Code (version 1.53.2)
[✓] Connected device (2 available)`

Was working fine before. Just encountered it.

**Comment by Mhamza-MM** [comment]
> Having same issue
> 
> Plugin `connectivity_macos` doesn't implement a plugin interface, nor sets a default implementation in pubspec.yaml.
> 
> To set a default implementation, use:
> flutter:
> plugin:
> platforms:
> macos:
> default_package:
> To implement an interface, use:
> flutter:
> plugin:
> implements:
> Plugin `path_provider_linux` doesn't implement a plugin interface, nor sets a default implementation in pubspec.yaml.
> 
> To set a default implementation, use:
> flutter:
> plugin:
> platforms:
> linux:
> default_package:
> 
> To implement an interface, use:
> flutter:
> plugin:
> implements:
> 
> Plugin `path_provider_windows` doesn't implement a plugin interface, nor sets a default implementation in pubspec.yaml.
> 
> To set a default implementation, use:
> flutter:
> plugin:
> platforms:
> windows:
> default_package:
> 
> To implement an interface, use:
> flutter:
> plugin:
> implements:
> 
> Plugin `shared_preferences_linux` doesn't implement a plugin interface, nor sets a default implementation in pubspec.yaml.
> 
> To set a default implementation, use:
> flutter:
> plugin:
> platforms:
> linux:
> default_package:
> 
> To implement an interface, use:
> flutter:
> plugin:
> implements:
> 
> Plugin `shared_preferences_macos` doesn't implement a plugin interface, nor sets a default implementation in pubspec.yaml.
> 
> To set a default implementation, use:
> flutter:
> 
> plugin:
> platforms:
> macos:
> default_package:
> 
> To implement an interface, use:
> flutter:
> plugin:
> implements:
> 
> Plugin `shared_preferences_windows` doesn't implement a plugin interface, nor sets a default implementation in pubspec.yaml.
> 
> To set a default implementation, use:
> flutter:
> plugin:
> platforms:
> windows:
> default_package:
> 
> To implement an interface, use:
> flutter:
> plugin:
> implements:
> 
> ../../../flutter/.pub-cache/hosted/pub.dartlang.org/flutter_svg-0.19.1/lib/src/picture_provider.dart:52:59: Error: No named parameter with the name 'nullOk'.
> context != null ? Localizations.localeOf(context, nullOk: true) : null,
> ^^^^^^
> ../../../flutter/packages/flutter/lib/src/widgets/localizations.dart:413:17: Context: Found this candidate, but the arguments don't match.
> 
> static Locale localeOf(BuildContext context) {
> ^^^^^^^^
> ../../../flutter/.pub-cache/hosted/pub.dartlang.org/win32-1.7.4+1/lib/src/structs.dart:1111:7: Error: Struct 'ENUMLOGFONTEX' is empty. Empty structs are undefined behavior.
> class ENUMLOGFONTEX extends Struct {
> ^
> ../../../flutter/.pub-cache/hosted/pub.dartlang.org/win32-1.7.4+1/lib/src/structs.dart:2835:7: Error: Struct 'BLUETOOTH_PIN_INFO' is empty. Empty structs are undefined behavior.
> class BLUETOOTH_PIN_INFO extends Struct {
> 
> ```
>   ^
> ```
> 
> ../../../flutter/.pub-cache/hosted/pub.dartlang.org/win32-1.7.4+1/lib/src/structs.dart:2960:7: Error: Struct 'EXCEPINFO' is empty. Empty structs are undefined behavior.
> class EXCEPINFO extends Struct {}
> ^
> ../../../flutter/.pub-cache/hosted/pub.dartlang.org/win32-1.7.4+1/lib/src/structs.dart:2966:7: Error: Struct 'PROPERTYKEY' is empty. Empty structs are undefined behavior.
> 
> class PROPERTYKEY extends Struct {}
> ^
> ../../../flutter/.pub-cache/hosted/pub.dartlang.org/win32-1.7.4+1/lib/src/structs.dart:2973:7: Error: Struct 'PROPVARIANT' is empty. Empty structs are undefined behavior.
> class PROPVARIANT extends Struct {}
> ^
> ../../../flutter/.pub-cache/hosted/pub.dartlang.org/win32-1.7.4+1/lib/src/structs.dart:2978:7: Error: Struct 'SAFEARRAY' is empty. Empty structs are undefined behavior.
> class SAFEARRAY extends Struct {}
> 
> ```
>   ^
> ```
> 
> ../../../flutter/.pub-cache/hosted/pub.dartlang.org/win32-1.7.4+1/lib/src/structs.dart:2985:7: Error: Struct 'CLSID' is empty. Empty structs are undefined behavior.
> 
> class CLSID extends Struct {}
> ^
> ../../../flutter/.pub-cache/hosted/pub.dartlang.org/win32-1.7.4+1/lib/src/structs.dart:2992:7: Error: Struct 'STATSTG' is empty. Empty structs are undefined behavior.
> class STATSTG extends Struct {}
> ^
> ../../../flutter/.pub-cache/hosted/pub.dartlang.org/win32-1.7.4+1/lib/src/structs.dart:2999:7: Error: Struct 'NLM_SIMULATED_PROFILE_INFO' is empty. Empty structs are undefined behavior.
> class NLM_SIMULATED_PROFILE_INFO extends Struct {}
> ^
> ../../../flutter/.pub-cache/hosted/pub.dartlang.org/ffi-0.1.3/lib/src/utf8.dart:23:7: Error: Struct 'Utf8' is empty. Empty structs are undefined behavior.
> class Utf8 extends Struct {
> ^
> 
> ../../../flutter/.pub-cache/hosted/pub.dartlang.org/ffi-0.1.3/lib/src/utf16.dart:16:7: Error: Struct 'Utf16' is empty. Empty structs are undefined behavior.
> class Utf16 extends Struct {
> 
> ```
>   ^
> ```
> 
> ../../../flutter/.pub-cache/hosted/pub.dartlang.org/ffi-0.1.3/lib/src/allocation.dart:47:33: Error: Expected type 'T' to be a valid and instantiated subtype of 'NativeType'.
> final int totalSize = count * sizeOf();
> ^
> 
> FAILURE: Build failed with an exception.
> 
> * Where:
>   Script '/Users/macbook/development/flutter/packages/flutter_tools/gradle/flutter.gradle' line: 991
> * What went wrong:
>   Execution failed for task ':app:compileFlutterBuildDebug'.
> 
> > Process 'command '/Users/macbook/development/flutter/bin/flutter'' finished with non-zero exit value 1
> 
> * Try:
>   Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.
> * Get more help at https://help.gradle.org
>   BUILD FAILED in 43s
>   Exception: Gradle task assembleDebug failed with exit code 1
>   Exited (sigterm)
> 
> flutter doctor ouput:
> 
> `Doctor summary (to see all details, run flutter doctor -v):
> 
> [✓] Flutter (Channel master, 1.27.0-5.0.pre.92, on macOS 11.2 20D64 darwin-x64, locale en-PK)
> 
> [✓] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
> 
> [✓] Xcode - develop for iOS and macOS
> 
> [✓] Chrome - develop for the web
> 
> [✓] Android Studio (version 4.1)
> 
> [✓] IntelliJ IDEA Ultimate Edition (version 2020.1)
> 
> [✓] VS Code (version 1.53.2)
> 
> [✓] Connected device (2 available)`
> 
> Was working fine before. Just encountered it.

So I Created This Project with Flutter stable channel and for some reasons I switched to master channel that is why the error was appearing. But it shouldn't. Then I switched back to stable channel after repairing pub cache(not sure if this step was required) and the error is gone.
`flutter pub cache repair`
`flutter channel`
`flutter upgrade`
The error is not appearing now and I can continue my work.
But I don't know the error was appearing with master channel. This should be resolved.

**Comment by jmagman** [comment]
@Mhamza-MM the ffi errors you see are unrelated to the non-blocking warnings this issue is tracking.  See https://github.com/flutter/flutter/issues/75172 and related issues.  If you believe this needs to be further investigated, please file a new issue.

**Comment by blasten** [comment]
This was fixed in #76561
